packages/runtime/src/runtime.ts

```diff
@@ -338,9 +338,9 @@ export class Runtime {
     const account = this.assertAccountDefined(accountAddr, this.store.accounts.get(accountAddr));
     if (appParams) {
       this.createOptInTx(accountAddr, appId, payFlags, flags);
-      this.run(program, ExecutionMode.STATEFUL); // execute TEAL code
-
       account.optInToApp(appId, appParams);
+
+      this.run(program, ExecutionMode.STATEFUL); // execute TEAL code
```

Comment:
Thanks for catching. Yes we should have a local state present before executing teal code. However, if the transaction fails, then the local state should not exist. 
To achieve this behaviour we can do a [`deepCopy`](https://medium.com/@manjuladube/understanding-deep-and-shallow-copy-in-javascript-13438bad941c) first (of the current state), then execute the transaction, and then finally update the 'real/actual' state with the copied one (if the tx execution passes successfully). Please let me know if that makes sense ;)
So, I would suggest to update the code of this function to:
```
  optInToApp (accountAddr: string, appId: number,
    flags: SSCOptionalFlags, payFlags: TxParams, program: string): void {
    const appParams = this.getApp(appId);
    if (appParams) {
      this.createOptInTx(accountAddr, appId, payFlags, flags);
      this.ctx.state = cloneDeep(this.store);
      const account = this.assertAccountDefined(
        accountAddr, this.ctx.state.accounts.get(accountAddr));
      account.optInToApp(appId, appParams);

      this.run(program, ExecutionMode.STATEFUL); // execute TEAL code
      this.store = this.ctx.state;
    } else {
      throw new TealError(ERRORS.TEAL.APP_NOT_FOUND, { appId: appId, line: 'unknown' });
    }
  }
```

Reply:
@ratik21 Thanks for the help, it makes sense. I've updated and rebased the PR.

Reply:
other idea would be to create a `context` object and use exactly same pattern we have when processing a transaction. @ratik21 - what do you think?

Reply:
Continued on discord.